dffml/util/file.py

```diff
@@ -148,3 +152,85 @@ def find_replace_with_hash_validation(
         replaced_path, expected_sha384_hash=expected_sha384_hash
     )
     return replaced_path
+
+
+class UnsupportedArchiveFormatError(Exception):
+    def __init__(self, format):
+        super().__init__()
+        self.format = format
+
+    def __str__(self):
+        return f"{self.format} format is not currently supported."
+
+
+class ArchiveHelper(ABC):
+    """
+    The base class for all archive helper classes
+    """
+
+    @abstractmethod
+    def __init__(self, path):
+        super().__init__()
+        # Path of the input file/folder
+        self.path = pathlib.Path(path)
+
+    @abstractmethod
+    def extract(self, output_dir=None):
+        """
+        Extracts the archive to output_dir 
+        If output_dir is None the file is extracted to the 
+        same parent directory as the file.
+        """
+
+    @abstractmethod
+    def add(self, output_path):
+        """
+        Adds the folder to an archive.
+        """
+
+    def fetch_file_paths(self, directory):
+        director_path = pathlib.Path(directory)
+        file_paths_list = [file_path for file_path in director_path.rglob("*")]
+        return file_paths_list
+
+
+class ZipHelper(ArchiveHelper):
+    def __init__(self, path):
+        super().__init__(path)
+
+    def extract(self, output_dir=None):
+        output_dir = self.path.parent if output_dir is None else output_dir
+        with ZipFile(self.path, "r") as zip:
+            zip.extractall(output_dir)
+
+    def add(self, output_path):
+        # TODO: handle the case for None output_path
+        file_path_list = self.fetch_file_paths(self.path)
+        with ZipFile(output_path, "w") as zip:
+            for file in file_path_list:
+                zip.write(file)
```

Comment:
Let's pick a few formats (stuff from standard library, zip, tar, etc.), and write functions / operations like the following.

Then when you go to implement within models. On load / save, look at the extension of the `location` property. Create two `DataFlow` objects where the operations in them are dependent the extension. One dataflow for loading from the location, one for saving back to the location.

**dffml/operations/archive.py**

```python
    @op
    def make_zip_archive(input_directory_path, output_path):
        with ZipFile(output_path, "w") as zip:
            for file in input_directory_path.rglob("*"):
                zip.write(file)

    @op
    def extract_zip_archive(input_path, output_directory_path):
        ...
```

Reply:
okay